File: test_src/fortran/hierarchical_parallelism/reduction/double_precision/target_parallel_loop.f90


program target_parallel_loop.f90


    

    implicit none
  
    INTEGER :: L = 5
    INTEGER :: i
    
    DOUBLE PRECISION :: COUNTER =  0   

    
    
     
    
    !$OMP TARGET PARALLEL LOOP   REDUCTION(+:COUNTER)   MAP(TOFROM: COUNTER) 


    DO i = 1 , L 


    

    


counter = counter +  1.  

 
     

    END DO

    !$OMP END TARGET PARALLEL LOOP
    

    IF  ( ( ABS(COUNTER - L) ) > 10*EPSILON( COUNTER   ) ) THEN
        write(*,*)  'Expected L Got', COUNTER
        call exit(1)
    ENDIF

end program target_parallel_loop.f90

File: test_src/cpp/hierarchical_parallelism/reduction/float/target__parallel__loop.cpp
#include <iostream>
#include <cmath>
#include <stdexcept>
bool almost_equal(float x, float gold, float tol) {
        return gold * (1-tol) <= x && x <= gold * ( 1+tol );
}
void test_target__parallel__loop(){
 // Input and Outputs
 const int L = 262144;
float counter{};
// Main program
#pragma omp target   map(tofrom:counter) 
{
#pragma omp parallel  reduction(+: counter)  
{
#pragma omp loop  
    for (int i = 0 ; i < L ; i++ )
{
counter += float { 1.0f };
    }
    }
    }
// Validation
if ( !almost_equal(counter,float { L }, 0.1)  ) {
    std::cerr << "Expected: " << L << " Got: " << counter << std::endl;
    throw std::runtime_error( "target__parallel__loop give incorect value when offloaded");
}
}
int main()
{
    test_target__parallel__loop();
}
